### PR TITLE
Lp extension

### PR DESCRIPTION
**Pull request by poliwop**
Adds the ability for LPs to have multiple LP positions in Omnipool in the same asset, by making LP positions NFTs. Backward compatible so that LP positions can still be represented by (pool_name, tkn_name) assets in agent holdings.

### CODE REVIEW

**Review by jepidoptera — CHANGES_REQUESTED** [review]
One higher-level critique: I don't know that it makes sense to have these two parallel systems for holding pool shares: nfts and the old system of ('omnipool', tkn). I think the nfts are a better and more flexible solution, and adding all this overhead to maintain backward compatibility seems worse in the long run than just refactoring things to use the new system. This works for now, but I propose simplifying in the next PR.

File: hydradx/model/amm/omnipool_amm.py [hunk 1]
```diff
@@ -1069,43 +1163,62 @@ def remove_liquidity(self, agent: Agent, quantity: float, tkn_remove: str):
                         f"Withdrawal rejected because the oracle volatility is too high: {volatility} > "
                         f"{self.remove_liquidity_volatility_threshold}", agent
                     )
-    
-        quantity = abs(quantity)
+
+        val = self.calculate_remove_liquidity(agent, quantity, tkn_remove, nft_id)
```
[the remark this answers is not in the document]
Comment (poliwop) [reply hydradx/model/amm/omnipool_amm.py:1167]: I agree, this was probably not intentional

File: hydradx/tests/test_omnipool_amm.py [hunk 2]
```diff
@@ -165,6 +166,81 @@ def test_add_liquidity_with_existing_position_fails(initial_state: oamm.Omnipool
...
+        tokens={
+            tkn: {'liquidity': liquidity[tkn], 'LRNA': lrna[tkn]} for tkn in lrna
+        }
+    )
+    tkn = initial_state.asset_list[0]
+    init_agent = Agent(holdings={tkn: initial_state.liquidity[tkn]})
+    delta_R = init_agent.holdings[tkn] / 2
+
+    single_add_state_1, single_add_agent_1 = oamm.simulate_add_liquidity(initial_state, init_agent, delta_R, tkn)
+    single_add_state_2, single_add_agent_2 = oamm.simulate_add_liquidity(initial_state, init_agent, delta_R, tkn,
```
[the remark this answers is not in the document]
Comment (poliwop) [reply hydradx/tests/test_omnipool_amm.py:214]: This is testing adding as an NFT to adding to holdings? I don't think that's redundant?

File: hydradx/model/amm/agents.py [hunk 3]
```diff
@@ -29,6 +30,7 @@ def __init__(self,
         self.trade_strategy = trade_strategy
         self.asset_list = list(self.holdings.keys())
         self.unique_id = unique_id
+        self.nfts = nfts or {}
 
     def __repr__(self):
         precision = 10
```
[the remark this answers is not in the document]
Comment (poliwop) [reply hydradx/model/amm/agents.py:36]: I'd prefer to address this in a future PR

**Review by jepidoptera — APPROVED** [review]

**Pull request merged**